error: relocation R_X86_64_TPOFF32 against hidden symbol `_ZN2v88internal18g_current_isolate_E' can not be used when making a shared object

When upgrading to the latest version of the `v8` create, I encounter the following error:
```
  = note: /usr/bin/ld: /tmp/rustcnxkAyd/libv8-fa1f4565c5216947.rlib(binding.o): relocation R_X86_64_TPOFF32 against hidden symbol `_ZN2v88internal18g_current_isolate_E' can not be used when making a shared object
          /usr/bin/ld: failed to set dynamic section sizes: bad value
          collect2: error: ld returned 1 exit status
```
despite the part of my code which references isolation being unchanged (see below)
```rust
        // let isolate_params = v8::CreateParams::default().heap_limits(0, 2000 * 1024 * 1024);
        let isolate = &mut v8::Isolate::new(Default::default());
        let handle_scope = &mut v8::HandleScope::new(isolate);
        let mut context = v8::Context::new(handle_scope, Default::default());
        let scope = &mut v8::ContextScope::new(handle_scope, context);
```
Do y'all have any pointers on how to debug/resolve this issue?

--- a/README.md
+++ b/README.md
@@ -157,6 +157,18 @@ is recommended.
 Arguments can be passed to `gn` by setting the `$GN_ARGS` environmental
 variable.
 
+For Linux targets, `rusty_v8` now defaults to defining
+`V8_TLS_USED_IN_LIBRARY` via GN args when building from source so the produced
+static archive can be linked into downstream `cdylib`/shared-library targets.
+The default injected argument is:
+
+```bash
+GN_ARGS='extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]'
+```
+
+Linux prebuilt release archives published by this repository are built with
+this shared-library-compatible TLS mode.
+
 Env vars used in when building from source: `SCCACHE`, `CCACHE`, `GN`, `NINJA`,
 `CLANG_BASE_PATH`, `GN_ARGS`
 

--- a/build.rs
+++ b/build.rs
@@ -346,10 +346,19 @@ fn build_v8(is_asan: bool) {
     println!("cargo:warning=Not using sccache or ccache");
   }
 
+  // Use the shared-library-safe TLS mode by default on Linux so downstream
+  // cdylibs can link rusty_v8 archives.
+  let mut has_tls_library_mode_define = false;
   if let Ok(args) = env::var("GN_ARGS") {
     for arg in args.split_whitespace() {
+      if arg.contains("V8_TLS_USED_IN_LIBRARY") {
+        has_tls_library_mode_define = true;
+      }
       gn_args.push(arg.to_string());
     }
+  }
+  if target_os == "linux" && !has_tls_library_mode_define {
+    gn_args.push(r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string());
   }
   // cross-compilation setup
   if target_arch == "aarch64" {

--- a/build.rs
+++ b/build.rs
@@ -1191,24 +1191,6 @@ fn env_bool(key: &str) -> bool {
 mod test {
   use super::*;
 
-  fn apply_linux_tls_logic_for_test(
-    gn_args: &mut Vec<String>,
-    target_os: &str,
-    raw_gn_args: Option<&str>,
-  ) {
-    let mut has_tls_library_mode_define = false;
-    if let Some(raw_gn_args) = raw_gn_args
-      && !raw_gn_args.trim().is_empty()
-    {
-      has_tls_library_mode_define =
-        raw_gn_args.contains("V8_TLS_USED_IN_LIBRARY");
-      gn_args.push(raw_gn_args.to_string());
-    }
-    if target_os == "linux" && !has_tls_library_mode_define {
-      gn_args.push(r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string());
-    }
-  }
-
   const MOCK_GRAPH: &str = r#"
 digraph ninja {
 rankdir="LR"
@@ -1256,61 +1238,5 @@ edge [fontsize=10]
     assert!(files.contains("../../../example/src/input.txt"));
     assert!(files.contains("../../../example/src/count_bytes.py"));
     assert!(!files.contains("obj/hello/hello.o"));
-  }
-
-  #[test]
-  fn test_linux_tls_override_detected_in_spaced_array() {
-    let mut gn_args = Vec::new();
-
-    apply_linux_tls_logic_for_test(
-      &mut gn_args,
-      "linux",
-      Some(r#"extra_cflags=[ "-O2", "-DV8_TLS_USED_IN_LIBRARY" ]"#),
-    );
-
-    assert_eq!(
-      gn_args,
-      vec![r#"extra_cflags=[ "-O2", "-DV8_TLS_USED_IN_LIBRARY" ]"#.to_string()]
-    );
-  }
-
-  #[test]
-  fn test_quoted_whitespace_gn_args_preserved_and_tls_default_injected() {
-    let mut gn_args = Vec::new();
-
-    apply_linux_tls_logic_for_test(
-      &mut gn_args,
-      "linux",
-      Some(r#"clang_base_path="/tmp/clang with spaces""#),
-    );
-
-    assert_eq!(
-      gn_args,
-      vec![
-        r#"clang_base_path="/tmp/clang with spaces""#.to_string(),
-        r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string(),
-      ]
-    );
-  }
-
-  #[test]
-  fn test_whitespace_only_gn_args_skipped() {
-    let mut gn_args = Vec::new();
-
-    apply_linux_tls_logic_for_test(&mut gn_args, "macos", Some("   "));
-
-    assert!(gn_args.is_empty());
-  }
-
-  #[test]
-  fn test_linux_tls_default_injected_without_override() {
-    let mut gn_args = Vec::new();
-
-    apply_linux_tls_logic_for_test(&mut gn_args, "linux", None);
-
-    assert_eq!(
-      gn_args,
-      vec![r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string()]
-    );
   }
 }

--- a/build.rs
+++ b/build.rs
@@ -347,16 +347,18 @@ fn build_v8(is_asan: bool) {
   }
 
   // Use the shared-library-safe TLS mode by default on Linux so downstream
-  // cdylibs can link rusty_v8 archives.
-  let mut has_tls_library_mode_define = false;
+  // cdylibs can link rusty_v8 archives. Skip injection when GN_ARGS already
+  // sets V8_TLS_USED_IN_LIBRARY.
+  let mut should_inject_tls_library_define = target_os == "linux";
   if let Ok(raw_gn_args) = env::var("GN_ARGS")
     && !raw_gn_args.trim().is_empty()
   {
-    has_tls_library_mode_define =
-      raw_gn_args.contains("V8_TLS_USED_IN_LIBRARY");
+    if raw_gn_args.contains("V8_TLS_USED_IN_LIBRARY") {
+      should_inject_tls_library_define = false;
+    }
     gn_args.push(raw_gn_args);
   }
-  if target_os == "linux" && !has_tls_library_mode_define {
+  if should_inject_tls_library_define {
     gn_args.push(r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string());
   }
   // cross-compilation setup

--- a/build.rs
+++ b/build.rs
@@ -349,16 +349,29 @@ fn build_v8(is_asan: bool) {
   // Use the shared-library-safe TLS mode by default on Linux so downstream
   // cdylibs can link rusty_v8 archives. Skip injection when GN_ARGS already
   // sets V8_TLS_USED_IN_LIBRARY.
-  let mut should_inject_tls_library_define = target_os == "linux";
+  let needs_tls_define = target_os == "linux";
+  let mut tls_define_injected = false;
   if let Ok(raw_gn_args) = env::var("GN_ARGS")
     && !raw_gn_args.trim().is_empty()
   {
     if raw_gn_args.contains("V8_TLS_USED_IN_LIBRARY") {
-      should_inject_tls_library_define = false;
+      tls_define_injected = true;
+      gn_args.push(raw_gn_args);
+    } else if needs_tls_define && raw_gn_args.contains("extra_cflags=[") {
+      // Prepend our define into the existing extra_cflags array so we don't
+      // silently override the user's flags with a second assignment.
+      let modified = raw_gn_args.replacen(
+        "extra_cflags=[",
+        r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY","#,
+        1,
+      );
+      tls_define_injected = true;
+      gn_args.push(modified);
+    } else {
+      gn_args.push(raw_gn_args);
     }
-    gn_args.push(raw_gn_args);
   }
-  if should_inject_tls_library_define {
+  if needs_tls_define && !tls_define_injected {
     gn_args.push(r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string());
   }
   // cross-compilation setup

--- a/build.rs
+++ b/build.rs
@@ -348,18 +348,12 @@ fn build_v8(is_asan: bool) {
 
   // Use the shared-library-safe TLS mode by default on Linux so downstream
   // cdylibs can link rusty_v8 archives.
-  let mut has_tls_library_mode_define = false;
-  if let Ok(args) = env::var("GN_ARGS") {
-    for arg in args.split_whitespace() {
-      if arg.contains("V8_TLS_USED_IN_LIBRARY") {
-        has_tls_library_mode_define = true;
-      }
-      gn_args.push(arg.to_string());
-    }
-  }
-  if target_os == "linux" && !has_tls_library_mode_define {
-    gn_args.push(r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string());
-  }
+  let raw_gn_args = env::var("GN_ARGS").ok();
+  add_user_gn_args_and_linux_tls_default(
+    &mut gn_args,
+    target_os,
+    raw_gn_args.as_deref(),
+  );
   // cross-compilation setup
   if target_arch == "aarch64" {
     gn_args.push(r#"target_cpu="arm64""#.to_string());
@@ -441,6 +435,24 @@ fn build_v8(is_asan: bool) {
     print_gn_args(&gn_out);
   }
   build("rusty_v8", None);
+}
+
+fn add_user_gn_args_and_linux_tls_default(
+  gn_args: &mut Vec<String>,
+  target_os: &str,
+  raw_gn_args: Option<&str>,
+) {
+  let raw_gn_args = raw_gn_args.filter(|args| !args.trim().is_empty());
+  let has_tls_library_mode_define =
+    raw_gn_args.is_some_and(|args| args.contains("V8_TLS_USED_IN_LIBRARY"));
+
+  if let Some(args) = raw_gn_args {
+    gn_args.push(args.to_string());
+  }
+
+  if target_os == "linux" && !has_tls_library_mode_define {
+    gn_args.push(r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string());
+  }
 }
 
 fn print_gn_args(gn_out_dir: &Path) {
@@ -1239,5 +1251,61 @@ edge [fontsize=10]
     assert!(files.contains("../../../example/src/input.txt"));
     assert!(files.contains("../../../example/src/count_bytes.py"));
     assert!(!files.contains("obj/hello/hello.o"));
+  }
+
+  #[test]
+  fn test_linux_tls_override_detected_in_spaced_array() {
+    let mut gn_args = Vec::new();
+
+    add_user_gn_args_and_linux_tls_default(
+      &mut gn_args,
+      "linux",
+      Some(r#"extra_cflags=[ "-O2", "-DV8_TLS_USED_IN_LIBRARY" ]"#),
+    );
+
+    assert_eq!(
+      gn_args,
+      vec![r#"extra_cflags=[ "-O2", "-DV8_TLS_USED_IN_LIBRARY" ]"#.to_string()]
+    );
+  }
+
+  #[test]
+  fn test_quoted_whitespace_gn_args_preserved_and_tls_default_injected() {
+    let mut gn_args = Vec::new();
+
+    add_user_gn_args_and_linux_tls_default(
+      &mut gn_args,
+      "linux",
+      Some(r#"clang_base_path="/tmp/clang with spaces""#),
+    );
+
+    assert_eq!(
+      gn_args,
+      vec![
+        r#"clang_base_path="/tmp/clang with spaces""#.to_string(),
+        r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string(),
+      ]
+    );
+  }
+
+  #[test]
+  fn test_whitespace_only_gn_args_skipped() {
+    let mut gn_args = Vec::new();
+
+    add_user_gn_args_and_linux_tls_default(&mut gn_args, "macos", Some("   "));
+
+    assert!(gn_args.is_empty());
+  }
+
+  #[test]
+  fn test_linux_tls_default_injected_without_override() {
+    let mut gn_args = Vec::new();
+
+    add_user_gn_args_and_linux_tls_default(&mut gn_args, "linux", None);
+
+    assert_eq!(
+      gn_args,
+      vec![r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string()]
+    );
   }
 }

--- a/build.rs
+++ b/build.rs
@@ -348,12 +348,17 @@ fn build_v8(is_asan: bool) {
 
   // Use the shared-library-safe TLS mode by default on Linux so downstream
   // cdylibs can link rusty_v8 archives.
-  let raw_gn_args = env::var("GN_ARGS").ok();
-  add_user_gn_args_and_linux_tls_default(
-    &mut gn_args,
-    target_os,
-    raw_gn_args.as_deref(),
-  );
+  let mut has_tls_library_mode_define = false;
+  if let Ok(raw_gn_args) = env::var("GN_ARGS")
+    && !raw_gn_args.trim().is_empty()
+  {
+    has_tls_library_mode_define =
+      raw_gn_args.contains("V8_TLS_USED_IN_LIBRARY");
+    gn_args.push(raw_gn_args);
+  }
+  if target_os == "linux" && !has_tls_library_mode_define {
+    gn_args.push(r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string());
+  }
   // cross-compilation setup
   if target_arch == "aarch64" {
     gn_args.push(r#"target_cpu="arm64""#.to_string());
@@ -435,24 +440,6 @@ fn build_v8(is_asan: bool) {
     print_gn_args(&gn_out);
   }
   build("rusty_v8", None);
-}
-
-fn add_user_gn_args_and_linux_tls_default(
-  gn_args: &mut Vec<String>,
-  target_os: &str,
-  raw_gn_args: Option<&str>,
-) {
-  let raw_gn_args = raw_gn_args.filter(|args| !args.trim().is_empty());
-  let has_tls_library_mode_define =
-    raw_gn_args.is_some_and(|args| args.contains("V8_TLS_USED_IN_LIBRARY"));
-
-  if let Some(args) = raw_gn_args {
-    gn_args.push(args.to_string());
-  }
-
-  if target_os == "linux" && !has_tls_library_mode_define {
-    gn_args.push(r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string());
-  }
 }
 
 fn print_gn_args(gn_out_dir: &Path) {
@@ -1204,6 +1191,24 @@ fn env_bool(key: &str) -> bool {
 mod test {
   use super::*;
 
+  fn apply_linux_tls_logic_for_test(
+    gn_args: &mut Vec<String>,
+    target_os: &str,
+    raw_gn_args: Option<&str>,
+  ) {
+    let mut has_tls_library_mode_define = false;
+    if let Some(raw_gn_args) = raw_gn_args
+      && !raw_gn_args.trim().is_empty()
+    {
+      has_tls_library_mode_define =
+        raw_gn_args.contains("V8_TLS_USED_IN_LIBRARY");
+      gn_args.push(raw_gn_args.to_string());
+    }
+    if target_os == "linux" && !has_tls_library_mode_define {
+      gn_args.push(r#"extra_cflags=["-DV8_TLS_USED_IN_LIBRARY"]"#.to_string());
+    }
+  }
+
   const MOCK_GRAPH: &str = r#"
 digraph ninja {
 rankdir="LR"
@@ -1257,7 +1262,7 @@ edge [fontsize=10]
   fn test_linux_tls_override_detected_in_spaced_array() {
     let mut gn_args = Vec::new();
 
-    add_user_gn_args_and_linux_tls_default(
+    apply_linux_tls_logic_for_test(
       &mut gn_args,
       "linux",
       Some(r#"extra_cflags=[ "-O2", "-DV8_TLS_USED_IN_LIBRARY" ]"#),
@@ -1273,7 +1278,7 @@ edge [fontsize=10]
   fn test_quoted_whitespace_gn_args_preserved_and_tls_default_injected() {
     let mut gn_args = Vec::new();
 
-    add_user_gn_args_and_linux_tls_default(
+    apply_linux_tls_logic_for_test(
       &mut gn_args,
       "linux",
       Some(r#"clang_base_path="/tmp/clang with spaces""#),
@@ -1292,7 +1297,7 @@ edge [fontsize=10]
   fn test_whitespace_only_gn_args_skipped() {
     let mut gn_args = Vec::new();
 
-    add_user_gn_args_and_linux_tls_default(&mut gn_args, "macos", Some("   "));
+    apply_linux_tls_logic_for_test(&mut gn_args, "macos", Some("   "));
 
     assert!(gn_args.is_empty());
   }
@@ -1301,7 +1306,7 @@ edge [fontsize=10]
   fn test_linux_tls_default_injected_without_override() {
     let mut gn_args = Vec::new();
 
-    add_user_gn_args_and_linux_tls_default(&mut gn_args, "linux", None);
+    apply_linux_tls_logic_for_test(&mut gn_args, "linux", None);
 
     assert_eq!(
       gn_args,

